lib/TcpConnection.js

```diff
@@ -92,23 +98,17 @@ module.exports = function factory (userOptions) {
   }
 
   function reconnect () {
-    const retry = backoff.fibonacci()
-    retry.failAfter(options.reconnectTries)
-    retry.on('ready', () => {
-      connect((err) => {
-        if (connected === false) return retry.backoff(err)
-      })
-    })
-    retry.on('fail', (err) => process.stderr.write(`could not reconnect: ${err.message}`))
-    retry.backoff()
+    retryBackoff.backoff()
   }
   // end: connection handlers
 
   // begin: connection listeners
   function closeListener (hadError) {
     disconnect()
     if (hadError) {
-      process.stderr.write(socketError.message)
+      if (options.onSocketClose) {
+        options.onSocketClose(socketError)
+      }
     }
     if (options.reconnect) reconnect()
```

Comment:
Here, we are using `reconnect` but we don't call `removeListeners()` (unlike the `endListener` function). Maybe the `disconnect` function should remove listeners? Alternatively, the `connect` could remove listeners on the current `socket` before creating a new connection?

Reply:
Either works, just remove the listeners :).

Reply:
Done!